File: mobile/lib/services/diagnostic_payload_codec.dart
import 'dart:convert';
import 'dart:io';

/// Compression gzip + base64 pour diagnostics et captures (stockage compact).
class DiagnosticPayloadCodec {
  DiagnosticPayloadCodec._();

  static const String compressedPrefix = 'gz:';

  static String compressJson(Map<String, dynamic> json) {
    final raw = utf8.encode(jsonEncode(json));
    final compressed = gzip.encode(raw);
    return '$compressedPrefix${base64Encode(compressed)}';
  }

  static String compressBytes(List<int> bytes) {
    final compressed = gzip.encode(bytes);
    return '$compressedPrefix${base64Encode(compressed)}';
  }

  static Map<String, dynamic> decompressJson(String payload) {
    final trimmed = payload.trim();
    if (!trimmed.startsWith(compressedPrefix)) {
      return jsonDecode(trimmed) as Map<String, dynamic>;
    }
    final raw = gzip.decode(base64Decode(trimmed.substring(compressedPrefix.length)));
    return jsonDecode(utf8.decode(raw)) as Map<String, dynamic>;
  }
}

File: backend/notification-service/src/utils/diagnosticPayloadCodec.js
const zlib = require('zlib');

const COMPRESSED_PREFIX = 'gz:';

function decompressJson(payload) {
  if (payload == null || typeof payload !== 'string') {
    throw new Error('Payload diagnostic invalide');
  }
  const trimmed = payload.trim();
  if (!trimmed.startsWith(COMPRESSED_PREFIX)) {
    return JSON.parse(trimmed);
  }
  const raw = zlib.gunzipSync(
    Buffer.from(trimmed.slice(COMPRESSED_PREFIX.length), 'base64'),
  );
  return JSON.parse(raw.toString('utf8'));
}

function tryDecompressJson(payload) {
  try {
    return decompressJson(payload);
  } catch {
    return null;
  }
}

function decompressBytes(payload) {
  if (payload == null || typeof payload !== 'string') {
    throw new Error('Payload binaire invalide');
  }
  const trimmed = payload.trim();
  if (!trimmed.startsWith(COMPRESSED_PREFIX)) {
    return Buffer.from(trimmed, 'base64');
  }
  return zlib.gunzipSync(
    Buffer.from(trimmed.slice(COMPRESSED_PREFIX.length), 'base64'),
  );
}

function tryDecompressBytes(payload) {
  try {
    return decompressBytes(payload);
  } catch {
    return null;
  }
}

module.exports = {
  decompressJson,
  tryDecompressJson,
  decompressBytes,
  tryDecompressBytes,
};
